### ISSUE TITLE
Allow to choose the variable (abundance, concentration, biovolume) for dwca export in API call

### ISSUE THREAD

**Issue by grololo06** [ecotaxa/ecotaxa_dev#630]
The dwca export can be amended in terms of abundances #626, concentrations #628, or biovolume #629.

Concentration is possible as a choice when the appropriate fields (total volume and subsampling coef) are specified in project settings (#619).
Biovolume is the same but needs individual volume to be specified too.

When abundance is chosen, sample and subsample are available.

When concentration or biovolume are available, sample or subsample is chosen automatically depending at which level the total volume is specified (cf #616 ).

**Comment by grololo06** [comment]
@jiho:
- if you have preferences for API syntax, feel free to tell
- Is it a single variable or several of them?
- What's the degraded mode when the projects composing the collection have different settings?

**Comment by jiho** [comment]
> * if you have preferences for API syntax, feel free to tell

Still reviewing the API for consistency. Time flies...
As general suggestions:
- export/dwca and export/full and export/summary look like good endpoints names
- if possible I'd be in favour of having 2 kinds of endpoints:
     -  those with only simple arguments (scalars): set all query parameters (i.e. no argument in the endpoint url itself) and an empty body
     - those with complex arguments : set them all in body and nothing as query params

> * Is it a single variable or several of them?

To determine the type of export? I'd say single variable with a vector argument. But make it as easy as possible for you.

> * What's the degraded mode when the projects composing the collection have different settings?

Concentration is only choosable if all underlying projects have total volume and subsampling coef set.
Same for biovolume.

Consistency is more important than volume of data, so the export should only provide something if it can be provided consistently for the whole dataset.

**Comment by grololo06** [comment]
To be crystal clear, the issue title is "Allow to choose the variable", is it rather "Allow to choose the variable**s**' ?

**Comment by jiho** [comment]
Indeed, I initially thought that it would be only one variable but:
- for dwca I suppose that several variables can be included in the eMoF table                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 - (for summary export, although this is not the purpose of this particular issue, it would be possible to have several variables if the data is formatted in "tall" format)
so several variables would be possible... unless it is too complicated to output.

**Comment by jiho** [comment]
If I may complicate this a bit further... 😕

Actually the choices should be 
- **occurrence: just presence**
- abundance: counts, including 0
- concentrations: count/m3, including 0
- biovolume: mm3/m3, including 0

The first possibility means nothing in the EMOF table and will be necessary for datasets that are not homogeneously sorted, for which abundances (and 0 abundances in particular) would be wrong.

**Comment by grololo06** [comment]
> Indeed, I initially thought that it would be only one variable but:
> 
>     * for dwca I suppose that several variables can be included in the eMoF table
> 
>     * for summary export (although this is not the purpose of this particular issue) it would be possible to have several variables if the data is formatted in "tall" format
> 
> 
> so several variables would be possible... unless it is too complicated to output.

Hum, let's assert the correctness (in v.2.6.8) of summary exports before extending them in a specific task/issue. I assume as well that several variables can be output in DwCA.

**Comment by grololo06** [comment]
Present state of the code:
- All DwCA events are EcoTaxa samples
- present-type occurrences are produced in occurence.txt, always
- absent-type occurrences are produced in occurence.txt, on-demand (API parameter 'with_zeroes')
- [count](http://vocab.nerc.ac.uk/collection/P01/current/OCOUNT01/) emof lines are produced, always
- [concentration](http://vocab.nerc.ac.uk/collection/P01/current/SDBIOL01/) emof lines are produced, with [hard-coded extraction formulae](https://github.com/ecotaxa/ecotaxa_back/blob/518efc25b8b4920d3b9c3ef063d629d8065dae6e/py/API_operations/exports/DarwinCore.py#L678), if input data is available and correct 
- [biovolume](http://vocab.nerc.ac.uk/collection/P01/current/CVOLUKNB/) emof lines are produced, with [hard-coded extraction formulae](https://github.com/ecotaxa/ecotaxa_back/blob/518efc25b8b4920d3b9c3ef063d629d8065dae6e/py/API_operations/exports/DarwinCore.py#L695), if input data is available and correct.

All occurrences and emofs are linked to an event, meaning the computations are for a sample.
 
What is planned already:
- The extraction formulae are from 2021 set of published collections, they will move to project-specific settings for a wider usage, it's possible in code since v2.6.8 but the data is not present yet.

**Comment by grololo06** [comment]
From API point-of-view, what can be done is:
- Add the choice**s** somehow e.g. several of {PRESENT_OCC, ABSENT_OCC, COUNT, CONCENTRATON, BIOVOLUME} as a parameter _and reject the call if a choice cannot be satisfied_ 
- Optionally, add a primitive to return the valid choices for a collection.

**Comment by grololo06** [comment]
Need to clear-up the part in bold:

> When abundance is chosen, **sample and subsample are available**.
> 
> When concentration or biovolume are available, **sample or subsample is chosen automatically** depending at which level the total volume is specified (cf #616 ).

Is it referring to the output format?

**Comment by jiho** [comment]
Text clarified in the first comment of this issue.

**Comment by juliecoust** [comment]
Déplacé sur ClickUp.
